### ISSUE TITLE
Support of destructuring assignment

### ISSUE THREAD

**Issue by oicu0619** [googleprojectzero/fuzzilli#231]
Hi, i am working on compile JS to Fuzzil IR. There is no operator in IR direct representing destructuring assignment, so i think there might be two ways to do the compilation:
1. transpile destructuring assignment to regular assignment then compile.
2. add some support from the Fuzzil side.

For 1, something like object destructing is simple, but array destructing, or object destructing with rest element is complicated. According to https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-runtime-semantics-destructuringassignmentevaluation or https://babeljs.io/repl#?browsers=ie%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=G4QwTgBA2iA0BGBdAvAWRAFwBYDoCmAHgA4C2AjABRkCUA3EA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.14.7&externalPlugins=  ,  array destructings should contain several iterable checks. Also V8 generates pretty long byte code for array destructuring. So i think it is not very easy to generator same javascript using regular assignment instead of destructuring assignment.

For 2, destructring assignment must be on the left hand side of some productions of grammar. It contains a set of variables. Also it support recursion and rest element such as 
```
var [[a,b],...c]=[[]];
```
I think it is quite different from the existing IR in some ways.

What is the property way to deal with it? Thanks a lot for the help!

**Comment by saelo** [comment]
Nice, yeah I agree with you that option (1) probably wouldn't be able to achieve the same coverage in the target engine, so at least a basic version of option (2) seems preferable. I could imagine an operation such as 

```swift
class Destruct: Operation {
    let format: X
    init(numOutputs: Int, format: X) {
        precondition(format.count == numOutput)     // or something like that
        self.format = format
        super.init(numInputs: 1, numOutputs: numOutputs)
    }
}
```
And possibly additional operations for destructuring objects (`const {a, b} = obj;`) and reassigning existing variables when desctructuring (e.g. `[a, b] = array;`).

Here `format` would somehow describe the nesting level of each output variable and whether it is a rest parameter (and ideally, already guarantee that only the last element of each nesting level can be a rest element). I think it'd also be fine to just do a simpler version of this first that supports no nesting but just has a `let hasRest: bool` field.

**Comment by oicu0619** [comment]
Thanks for the reply!
I think the simpler version of Destruct Operator is quite expressive. Something nesting like ```[v0,{b:v1},v2] =array``` can be compile to multiple destruct Operator:
```
v0, v3, v2<- ArrayDestruct array
v1<- ObjectDestruct v3
```
i think that they are almost the same thing to the JIT engine. 

Also, do you think we need both Destruct and ReassignDestruct, or we declare variables via LoadUndefined, and only keep ReassignDestruct? The latter one is one IR OP less but introduce additional Reassign in the IR program.

Also, i have a question about compiling JS to IR, about hoisting. 
1. need to hoist VAR declaration to the top of the scope.
2. have to hoist function declaration to the top of the function scope, because everything should be declared before use.
3. have to change function declaration to function expression```v1=undefined;v2=function(){}; v0=v2``` to solve the mutual recursive problem.
4. have to hoist every LET/CONST declaration because function is hoisted, the variable may be used in the function body.
This means almost everything is hoisted to the top of the scope and create a lot of LoadUndefined and Reassigns. 
Does unnecessary Reassign affect further mutation of anything else? Does it worth it to keep in track and only hoist(Reassign) the needed variables/functions?

Thanks for the help!

**Comment by saelo** [comment]
Nice, yeah going with just the simpler version sounds good to me!

I think I'd go with just a `Destruct` operation at first (so the variant that creates new variables), and then possibly a `DestructReassign` later, but that's probably not too important (?). Otherwise, as you said there'd probably be a lot of `LoadUndefined` instruction, which is kinda weird/inefficient.

Re. hoisting:
Yes, I think it would make sense to only hoist the necessary variables and functions, to keep the total number of instructions and variables low (which then increases mutation efficiency).
But yeah, the hoisting issue isn't great. It's probably also a problem for code like this:

```javascript
function foo() {
    print(greeting);
}
let greeting = "Hello World!";
foo();
```

which wouldn't really work in FuzzIL?

Maybe there's something we can do about this in FuzzIL, to make it easier to compiler such constructs. One option might be to use [named variables](https://github.com/googleprojectzero/fuzzilli/issues/224), although then you'd probably have to turn all variables in a JS file into such FuzzIL NamedVariables. But maybe that's fine! Let me know what you think

**Comment by oicu0619** [comment]
Sure. I will try to send a PR about ```ArrayDestruct``` and ```ObjectDestruct``` . 

About Hoisting:
for the code snippet, i used to just hoist the let variable greeting then compile, like
```
let greeting;
function foo() {
    print(greeting);
}
greeting = "Hello World!";
foo();
```
but this is not semantically same. 

Named variables seems can solved the problem. I don't know whether i understand it correctly, should i compile the code snippet to something like:
```
v0 <- BeginPlainFunctionDefinition ->
    LoadNamedVariable 'greeting' V1
    LoadNamedVariable 'print' V2
    V3 <- CallFunction V2 V1
EndPlainFunctionDefinition
StoreNamedVariable 'foo' V0
V4 <- LoadString 'Hello World!'
StoreNamedVariable 'greetings' V4
LoadNamedVariable 'foo' V5
CallFunction V5
```
But I think  named variables generate one additional IR Operation for every variable declaration/load/store ,which is more IR Operation then hoisting. Hoisting generate additional IR for every declaration (LoadUndefined). Do you think the additional IR is OK? Thanks!

**Comment by saelo** [comment]
Yeah I think that's pretty much how it would look like, and yes, this requires more IR operations than hoisting, and so probably should still only be done when necessary and not always. Also, the JavaScript lifter then probably needs to track which named variables are already defined in the current scope, to know when to emit a `let` and when not to. On the other hand, this is probably pretty close to how a JS engine actually handles these things (like, they'll also have IR operations to look up a named variable), so that's probably a good thing. I think if it's useful/necessary for your use case, then we should add this to FuzzIL, in the absence of any better alternatives (I can't currently think of anything better).

**Comment by saelo** [comment]
We now have named variables: https://github.com/googleprojectzero/fuzzilli/commit/f7dfd455969c24a31af182fafde889c0ff68ebac which can be used when hoisting is required. We could probably also introduce something like a "named function" to deal with function hoisting (or alternatively assign the function to a named variable). I'll close this issue for now, but feel free to reopen if it's still relevant!
